Create a dummy sandbox when the sandbox config can't be satisfied

abderrahim opened the pull request:
Instead of an error, a dummy sandbox can be used for checking out an
artifact for a different OS/arch if integration commands don't need to
be run.

Should fix #1352 (not tested yet)

abderrahim committed:
platform: merge check_sandbox_config into create_sandbox

It's only called just before creating a sandbox

jjardon commented on .github/workflows/ci.yml:
```diff
@@ -34,8 +34,8 @@ jobs:
-      CI_IMAGE_SUFFIX: master-177137613
-      TOXENV: py36,py37,py38-nocover,py36-plugins,py37-plugins,py38-plugins-nocover
+      CI_IMAGE_SUFFIX: master-236878594
+      TOXENV: py36,py37,py38-nocover,py36-plugins,py37-plugins,py38-plugins-nocover,py39
```
Sure, done

jjardon replied:
mmm, I think it passes the CI in my fork with py39 only; let me use nocover for now and maybe change that in another PR later on

jjardon commented on .github/workflows/ci.yml:
```diff
@@ -34,8 +34,8 @@ jobs:
       CI_IMAGE_PREFIX: registry.gitlab.com/buildstream/buildstream-docker-images/testsuite
-      CI_IMAGE_SUFFIX: master-177137613
-      TOXENV: py36,py37,py38-nocover,py36-plugins,py37-plugins,py38-plugins-nocover
+      CI_IMAGE_SUFFIX: master-236878594
```
Sure, done

jjardon commented:
Ci in my fork: https://github.com/jjardon/buildstream/actions/runs/467133628

abderrahim committed:
platform: return a dummy sandbox when the sandbox config can't be satisfied

Instead of an error, a dummy sandbox can be used for checking out an
artifact for a different OS/arch if integration commands don't need to
be run.

Fixes #1352

nanonyme commented on src/buildstream/_platform/platform.py:
```diff
@@ -159,44 +161,16 @@ def get_host_arch():
+            dummy_reasons = " and ".join(self.dummy_reasons)
+        else:
+            try:
+                SandboxBuildBoxRun.check_sandbox_config(kwargs['config'])
```
Looks like linter wants you to use double quotes here.

nanonyme commented on src/buildstream/_platform/platform.py:
```diff
@@ -159,44 +161,16 @@ def get_host_arch():
+                SandboxBuildBoxRun.check_sandbox_config(kwargs['config'])
+                return SandboxBuildBoxRun(*args, **kwargs)
+            except SandboxError as e:
+                dummy_reasons = e.message
```
Should probably be `dummy_reasons = str(e)`. message is passed to Exception() as only parameter, not stored as member. (see https://github.com/apache/buildstream/blob/c2ae6d47aad0686d29d34208e83f0447f5c4591b/src/buildstream/_exceptions.py#L93)

gtristan replied:
Yup, even if this is simply stylistic (it *might* be `e.message` ?), I agree we should use `str(e)` here for consistency.

nanonyme replied:
Well, I looked into the inheritance hierarchy and it looked like it really cannot ever be e.message. Unless you have some class which *inherits* SandboxError and adds message as member. But SandboxError and its base classes do not do that so linter was right. Perhaps it's a bug and they *should* set it as member.

gtristan commented:
I'm not particularly fond of this patch and how the `Platform` is digging quite deep into the `Sandbox` interfaces instead of letting `Sandbox` handle that bit, although this seems to already be a bit messy before the patch.

Looking at this, my main concern is that we are no longer checking the sandbox config when creating a sandbox in `Element`, so this could cause some sort of regression in error reporting when actually trying to run commands in the sandbox where the config is not supported.

It seems to me (and I might very well be missing something) that we need to have some semantic in `Element.__sandbox()` to dictate whether we need the sandbox for the purpose of *running commands* or not, in order to determine if we need to run this check or not.

abderrahim commented:
> I'm not particularly fond of this patch and how the Platform is digging quite deep into the Sandbox interfaces instead of letting Sandbox handle that bit, although this seems to already be a bit messy before the patch.

Yeah, it is messy. I want to refactor it, but I'd rather have a small patch that can be backported to bst-1 at first.

> Looking at this, my main concern is that we are no longer checking the sandbox config when creating a sandbox in Element, so this could cause some sort of regression in error reporting when actually trying to run commands in the sandbox where the config is not supported.

Checking the sandbox config was merged into `create_sandbox`, since it was only ever called just before creating the sandbox (this is the first commit).

> It seems to me (and I might very well be missing something) that we need to have some semantic in Element.__sandbox() to dictate whether we need the sandbox for the purpose of running commands or not, in order to determine if we need to run this check or not.

What this MR does is to make the check not raise an error, but instead return a dummy sandbox which will raise an error when something tries to use it. This is what the old code was supposed to do, but got broken at some point.

abderrahim commented:
@cs-shadow @gtristan OK to merge this? I'd like to backport it to 1.6 too.

abderrahim commented:
I'm going to merge this since nobody wants to review it.
